### DISCUSSION TITLE
Is it able to get request context in hook?

### DISCUSSION THREAD

**Discussion by goxiaoy** [directus/directus#5621] [Archive]
I have put some custom header into request header, and want to extract it in the custom hook function. I have tried to use [express-http-context](https://www.npmjs.com/package/express-http-context), but it seems not work. Any idea?

**Comment by rijkvanzanten** [comment]
> and want to extract it in the custom hook function

What hook are you using? The `request` hook lets you access the whole request. Item-modifier hooks (like items.create) etc don't have access to the request, as it's not a given they're always used from an HTTP endpoint.

**Reply by goxiaoy** [reply]
> Item-modifier hooks (like items.create) etc don't have access to the request, as it's not a given they're always used from an HTTP endpoint.

Yes, I am using items related hooks and I know these events may be triggered outside http clearly.
But in my case, I would like to record some http related information before item being created, like ip address, user's timezone. So is there a way to access the http related info if it happens in a http request?

**Reply by rijkvanzanten** [reply]
No, you don't have access to the HTTP request in the item hook. That being said, you should be able to create a custom endpoint that combines the object payload with the request data, and then insert that using the regular items service creation flow 👍🏻

**Reply by goxiaoy** [reply]
This does not work because I am using GraphQL and there are so many things to rewrite

**Comment by goxiaoy** [comment]
Oh I figure it out
1. create a custom hook which put some value into httpcontext
``` javascript
    return {
        "middlewares.init.before": async function ({ app }: { app: Express }) {
            app.use(httpContext.middleware)
            app.use((req, res, next) => {
                // https://stackoverflow.com/a/55995352/15527691
                httpContext.ns.bindEmitter(req);
                httpContext.ns.bindEmitter(res);
                httpContext.set(req, req.get('X-TZ'))
                next();
            });
        }
    };
```

2. In your items hook
``` javascript
    return {
        'items.create.before': async function (input, { collection, payload }) {
           console.log(httpContext.get('tz'))
        }
    }
```



ref: 
https://www.npmjs.com/package/express-http-context

**Reply by rijkvanzanten** [reply]
Just be aware of any timing issues with that. The request called in the first handler might not be the same that's causing the items.create hook
